SignalMessaging/attachments/ApproveContactShareViewController.swift

```diff
@@ -0,0 +1,556 @@
+//
+//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+import SignalServiceKit
+
+@objc
+public protocol ApproveContactShareViewControllerDelegate: class {
+    func approveContactShare(_ approveContactShare: ApproveContactShareViewController, didApproveContactShare contactShare: OWSContact)
+    func approveContactShare(_ approveContactShare: ApproveContactShareViewController, didCancelContactShare contactShare: OWSContact)
+}
+
+// MARK: -
+
+class ContactShareField: NSObject {
+
+    private var isIncludedFlag = true
+
+    func localizedLabel() -> String {
+        preconditionFailure("This method must be overridden")
+    }
+
+    func isIncluded() -> Bool {
+        return isIncludedFlag
+    }
+
+    func setIsIncluded(_ isIncluded: Bool) {
+        isIncludedFlag = isIncluded
+    }
+
+    func applyToContact(contact: OWSContact) {
+        preconditionFailure("This method must be overridden")
+    }
+}
+
+// MARK: -
+
+class ContactSharePhoneNumber: ContactShareField {
+
+    let value: OWSContactPhoneNumber
+
+    required init(_ value: OWSContactPhoneNumber) {
+        self.value = value
+
+        super.init()
+    }
+
+    override func localizedLabel() -> String {
+        return value.localizedLabel()
+    }
+
+    override func applyToContact(contact: OWSContact) {
+        assert(isIncluded())
+
+        var values = [OWSContactPhoneNumber]()
+        values += contact.phoneNumbers
+        values.append(value)
+        contact.phoneNumbers = values
+    }
+}
+
+// MARK: -
+
+class ContactShareEmail: ContactShareField {
+
+    let value: OWSContactEmail
+
+    required init(_ value: OWSContactEmail) {
+        self.value = value
+
+        super.init()
+    }
+
+    override func localizedLabel() -> String {
+        return value.localizedLabel()
+    }
+
+    override func applyToContact(contact: OWSContact) {
+        assert(isIncluded())
+
+        var values = [OWSContactEmail]()
+        values += contact.emails
+        values.append(value)
+        contact.emails = values
+    }
+}
+
+// MARK: -
+
+class ContactShareAddress: ContactShareField {
```

Comment:
It's not a big deal, but it seems like making ContactShareField generic over the `value` would remove some redundancy in these field classes, so ultimately you could write something like:

```     
class ContactShareAddress: ContactShareField<OWSContactAddress>
```

I can follow up with an example later if you'd like.

Reply:
I've done it, but I think it's marginally worse.